File: Blind75/Graphs/Graph Valid Tree.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        # For a graph to be a tree:
        # 1. It must be fully connected (all nodes reachable)
        # 2. It must have no cycles
        # 3. It must have exactly n - 1 edges
        if len(edges) != n - 1:
            return False

        # Build adjacency list representation of the graph
        adj = [[] for _ in range(n)]
        for a, b in edges:
            adj[a].append(b)
            adj[b].append(a)

        visit = set()  # to keep track of visited nodes during DFS

        def dfs(node, parent):
            """
            Returns False if a cycle is detected, otherwise True.
            'parent' is used to avoid falsely detecting a cycle
            through the immediate parent in an undirected graph.
            """
            if node in visit:
                return False  # visiting an already visited node → cycle detected

            visit.add(node)

            # Explore all adjacent (neighboring) nodes
            for nei in adj[node]:
                if nei == parent:
                    continue  # skip the parent to prevent false cycle detection
                if not dfs(nei, node):
                    return False  # cycle detected deeper in recursion

            return True  # no cycle found from this node

        # Start DFS from node 0 (graph is assumed to have nodes 0..n-1)
        # If a cycle is found, return False immediately
        if not dfs(0, -1):
            return False

        # After DFS, check if all nodes were visited (ensures connectivity)
        return len(visit) == n

File: NeedCode 150 (JAVA)/Graphs/Graph Valid Tree.java
class Solution {
    int[] parent;
    int[] rank;

    public boolean validTree(int n, int[][] edges) {
        // A valid tree must have exactly n - 1 edges
        if (edges.length != n - 1) return false;

        parent = new int[n];
        rank = new int[n];

        for (int i = 0; i < n; i++) {
            parent[i] = i;
            rank[i] = 1;
        }

        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];

            int pu = find(u);
            int pv = find(v);

            // Cycle detected
            if (pu == pv) return false;

            union(pu, pv);
        }

        return true;
    }

    private int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    private void union(int x, int y) {
        if (rank[x] < rank[y]) {
            parent[x] = y;
        } else if (rank[x] > rank[y]) {
            parent[y] = x;
        } else {
            parent[y] = x;
            rank[x]++;
        }
    }
}
